src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -437,51 +457,121 @@ public void delete(String modelId, ActionListener<DeleteModelResponse> listener)
+                    )
                 );
             }, e -> listener.onResponse(new DeleteModelResponse(modelId, "failed", e.getMessage())));
 
-            // On model metadata removal, delete the model from the index
-            ActionListener<AcknowledgedResponse> onMetadataUpdateListener = ActionListener.wrap(
-                acknowledgedResponse -> deleteRequestBuilder.execute(onModelDeleteListener),
-                listener::onFailure
-            );
+            clearModelFromCacheStep.whenComplete(removeModelFromCacheResponse -> {
+                // After clearing the cache, if there are no errors remove modelId from blocked list
+                if (!removeModelFromCacheResponse.hasFailures()) {
+                    client.execute(
+                        UpdateBlockedModelAction.INSTANCE,
+                        new UpdateBlockedModelRequest(modelId, true),
+                        ActionListener.wrap(unblockModelIdStep::onResponse, unblockModelIdStep::onFailure)
+                    );
+                } else {
+                    String failureMessage = buildRemoveModelErrorMessage(modelId, removeModelFromCacheResponse);
+                    listener.onResponse(new DeleteModelResponse(modelId, "failed", failureMessage));
+                }
+            }, e -> listener.onResponse(new DeleteModelResponse(modelId, "failed", e.getMessage())));
 
-            // Remove the metadata asynchronously
+            // After unblocking modelId return the response
+            unblockModelIdStep.whenComplete(acknowledgedResponse -> {
+                listener.onResponse(new DeleteModelResponse(modelId, "deleted", null));
+                return;
+            }, listener::onFailure);
+        }
+
+        // This function helps to remove the model from blocked list when the delete request fails
+        // while executing after adding modelId to blocked list
+        private void unblockModelIdOnFailure(String modelId, Exception exceptionFromPreviousStep, StepListener<?> step) {
+            // If modelId is unblocked successfully, then we will just return the exception received from failed stepListener
+            // If unblocking modelId request fails, then we will return this exception along with the one received from failed stepListener
             client.execute(
-                UpdateModelMetadataAction.INSTANCE,
-                new UpdateModelMetadataRequest(modelId, true, null),
-                onMetadataUpdateListener
+                UpdateBlockedModelAction.INSTANCE,
+                new UpdateBlockedModelRequest(modelId, true),
+                ActionListener.wrap(acknowledgedResponse -> step.onFailure(exceptionFromPreviousStep), unblockingFailedException -> {
+                    String errorMsg = exceptionFromPreviousStep.getMessage();
+                    errorMsg = errorMsg + "\n" + unblockingFailedException.getMessage();
```

Comment:
we never used it. I have updated it to just return exception from previous step. I and Jack thought that may be it is not required to return unblocking failed exception because user may not be able to block or unblock a model explicitly.

Reply:
Now, we are not concatenating both error messages. We are just returning the exception returned from previous stepListener